### PR TITLE
[VP][RISCV] Introduce llvm.vp.minimum/maximum intrinsics

### PR DESCRIPTION
**Pull request by simeonkr**
Although there are predicated versions of minnum/maxnum, the ones for minimum/maximum are currently missing. This patch introduces these intrinsics and implements their lowering to RISC-V.

### CODE REVIEW

**Comment by simeonkr** [comment]
Ping

**Review by arsenm — APPROVED** [review]

**Comment by wangpc-pp** [comment]
The title should be `[VP]` instead of `[LV]`. This patch is not about Loop Vectorizer.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -6605,6 +6617,13 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
          !Subtarget.hasVInstructionsF16()))
       return SplitVPOp(Op, DAG);
     return lowerVectorFTRUNC_FCEIL_FFLOOR_FROUND(Op, DAG, Subtarget);
+  case ISD::VP_FMAXIMUM:
+  case ISD::VP_FMINIMUM:
+    if (Op.getValueType() == MVT::nxv32f16 &&
+        (Subtarget.hasVInstructionsF16Minimal() &&
+         !Subtarget.hasVInstructionsF16()))
+      return SplitVPOp(Op, DAG);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:6625]: Is this if tested. I didn't see any nxv32f16 test cases.

Comment (simeonkr) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:6625]: I've added them now.

File: llvm/test/CodeGen/RISCV/rvv/fmaximum-vp.ll [hunk 2]
```diff
@@ -0,0 +1,1245 @@
...
+;
+; ZVFHMIN-LABEL: vfmax_vv_nxv1f16:
+; ZVFHMIN:       # %bb.0:
+; ZVFHMIN-NEXT:    vmv1r.v v10, v0
+; ZVFHMIN-NEXT:    vsetvli a1, zero, e16, mf4, ta, ma
+; ZVFHMIN-NEXT:    vfwcvt.f.f.v v11, v8
+; ZVFHMIN-NEXT:    vsetvli zero, a0, e32, mf2, ta, ma
+; ZVFHMIN-NEXT:    vmfeq.vv v0, v11, v11, v0.t
+; ZVFHMIN-NEXT:    vsetvli a1, zero, e16, mf4, ta, ma
+; ZVFHMIN-NEXT:    vfwcvt.f.f.v v8, v9
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fmaximum-vp.ll:35]: This is a VLMAX vfwcvt.f.f.v, it should use the VL from the VP intrinsic.

Comment (simeonkr) [reply llvm/test/CodeGen/RISCV/rvv/fmaximum-vp.ll:35]: This seems to be the case for `llvm.vp.*` intrinsics in general (for example, as seen in `llvm/test/CodeGen/RISCV/rvv/vfmax-vp.ll` and `llvm/test/CodeGen/RISCV/rvv/vfadd-vp.ll`). Should this be fixed everywhere (in a separate patch)?

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fmaximum-vp.ll:35]: Yes that should be fixed. Extra changes of VL should be avoided as much as possible.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by simeonkr** [comment]
Can someone with commit access please merge this? Thank you.

**Comment by arsenm** [comment]
> Can someone with commit access please merge this? Thank you.

The bot shows a failure in   BOLT :: RISCV/relax.s, is that unrelated?

**Comment by simeonkr** [comment]
> The bot shows a failure in BOLT :: RISCV/relax.s, is that unrelated?

Yes, it's unrelated. It's passing after I rebased.

**Pull request merged**